openedx/frontend-app-authn
# feat: redirect to custom URL when third-party auth account is unlinked

**Settings**

```yaml
TUTOR_GROVE_MFE_LMS_COMMON_SETTINGS:
  MFE_CONFIG["TPA_UNLINKED_ACCOUNT_PROVISION_URL"] = "http://example.com/"

TUTOR_GROVE_LMS_ENV_FEATURES: |
  ENABLE_THIRD_PARTY_AUTH: True
  ENABLE_COMBINED_LOGIN_REGISTRATION: True
```
### Description

Allow redirecting to a custom URL when signing in via third-party auth when the account is not linked.

#### How Has This Been Tested?

1. Set the following settings in the LMS:
   - `ENABLE_MFE_CONFIG_API = True`
   - `ENABLE_COMBINED_LOGIN_REGISTRATION = True`
   - `ENABLE_THIRD_PARTY_AUTH = True`
   - `AUTHN_MICROFRONTEND_URL = 'http://localhost:1999'`
   - `MFE_CONFIG = {"TPA_UNLINKED_ACCOUNT_PROVISION_URL": "http://example.com"}`
2. Set `'ENABLE_THIRD_PARTY_AUTH': True` and `'ENABLE_AUTHN_MICROFRONTEND': True` in FEATURES dict in LMS settings (`edx-platform/lms/envs/common.py`).
3. Create a dummy backend Oauth2 provider at http://localhost:18000/admin/third_party_auth/oauth2providerconfig/ and enable its "Visible" setting.
4. Set `MFE_CONFIG_API_URL='http://localhost:18000/api/mfe_config/v1'` in frontend-app-authn (e.g. in `.env.development`).
5. Go to http://localhost:1999/login.
6. Sign in with the "Dummy" provider.
7. See that you are redirected to example.com

The login page will briefly show before redirect but I see no way to prevent this as the request to http://localhost:18000/api/mfe_context is made asynchronously and we don't know the third-party auth context until it is made, but the login page can be rendered before that.

#### Merge Checklist

* [ ] If your update includes visual changes, have they been reviewed by a designer? Send them a link to the Sandbox, if applicable. - N/A

#### Post-merge Checklist

* [ ] Deploy the changes to prod after verifying on stage or ask **@openedx/vanguards** to do it.
